SU2_CFD/src/output/filewriter/CSTLFileWriter.cpp

```diff
@@ -0,0 +1,346 @@
+   the writing of the STL file after gathering all of the data. ---*/
+
+  nLocalTria = dataSorter->GetnElem(TRIANGLE);
+  nLocalQuad = dataSorter->GetnElem(QUADRILATERAL);
+  nLocalTriaAll = nLocalTria + nLocalQuad*2; // Quad splitted into 2 tris
+
+  if (rank == MASTER_NODE) Buffer_Recv_nTriaAll = new unsigned long[nProcessor];
+
+  /*--- Communicate the maximum of local triangles on any process to each partition and the number of local vertices on each partition
+   to the master node with collective calls. ---*/
+
+  SU2_MPI::Allreduce(&nLocalTriaAll, &MaxLocalTriaAll, 1,
+                     MPI_UNSIGNED_LONG, MPI_MAX, MPI_COMM_WORLD);
+
+  SU2_MPI::Gather(&nLocalTriaAll,       1, MPI_UNSIGNED_LONG,
+                  Buffer_Recv_nTriaAll, 1, MPI_UNSIGNED_LONG,
+                  MASTER_NODE, MPI_COMM_WORLD);
+
+  /*--- Allocate buffer for send/recv of the coordinate data. Only the master rank allocates buffers for the recv. ---*/
+  bufD_Send = new su2double[MaxLocalTriaAll*3*3](); /* Triangle has 3 Points with 3 coords each */
+  if (rank == MASTER_NODE)
+    bufD_Recv = new su2double[nProcessor*MaxLocalTriaAll*3*3];
+
+  /*--- Load send buffers with the local data on this rank. ---*/
+  /*--- Write triangle element coordinate data into send buffer---*/
+  unsigned long index = 0;
+  for (iElem = 0; iElem < nLocalTria; iElem++) {
+    for (iPoint = 0; iPoint < 3; iPoint++) {
+
+      global_node_number = dataSorter->GetElem_Connectivity(TRIANGLE, iElem, iPoint) - 1; // (var, GlobalPointindex)
+      unsigned long local_node_number = global_node_number - dataSorter->GetNodeBegin(rank);
+
+      for (iVar = 0; iVar < 3; iVar++){
+        if (dataSorter->FindProcessor(global_node_number) == rank) {
+          bufD_Send[index] = dataSorter->GetData(iVar, local_node_number);
+        } else {
+          bufD_Send[index] = GetHaloNodeValue(global_node_number, iVar);
+        }
+
+        index++;
+      }//iVar
+    }//iPoint
+  }//iElem
+
+  /*--Write quadrilateral element coordinate data into send buffer:
+    Each quad is split into two triangles after a fixed node order. ---*/
+  for (iElem = 0; iElem < nLocalQuad; iElem++) {
+    for (iPoint = 0; iPoint < Nodelist.size(); iPoint++) {
+      global_node_number = dataSorter->GetElem_Connectivity(QUADRILATERAL,iElem,Nodelist[iPoint]) - 1; // (var, GlobalPointindex)
+      unsigned long local_node_number = global_node_number - dataSorter->GetNodeBegin(rank);
+
+      for (iVar = 0; iVar < 3; iVar++){
+        if (dataSorter->FindProcessor(global_node_number) == rank) {
+          bufD_Send[index] = dataSorter->GetData(iVar, local_node_number);
+        } else {
+          bufD_Send[index] = GetHaloNodeValue(global_node_number, iVar);
+        }
+
+        index++;
+      }//iVar
+    }//iPoint
+  }//iElem
+
+  /*--- Collective comms of the solution data and global IDs. ---*/
+  SU2_MPI::Gather(bufD_Send, static_cast<int>(MaxLocalTriaAll*3*3), MPI_DOUBLE,

```

Comment:
Use N_POINTS_TRIANGLE, etc., to be more clear (this is in a few places)

Reply:
Do you mean replacing `MaxLocalTriaAll*3*3 ` by `N_POINTS_TRIANGLE` ? This is actually a coordinate (9 per triangle) count. I introduced `max_nLocalCoords = MaxLocalTriaAll*3*3` -> Does that fully address your point?

Reply:
I think Tom is referring to the number 3 being replaced by the constant defined in option_structure.
Of course you have two threes xD, one for points the other for dimensions (i.e. don't do  N_POINTS_TRIANGLE*N_POINTS_TRIANGLE).

Reply:
👍

Reply:
🤦‍♂